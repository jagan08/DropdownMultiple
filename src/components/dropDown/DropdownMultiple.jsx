import React, {Component} from 'react';
import FontAwesome from 'react-fontawesome';
import './dropDownMultiple.css';
import Searchfilter from '../searchFilter/Searchfilter';
import CheckboxMultipleSelect from '../checkBox/CheckboxMultipleSelect';

class DropdownMultiple extends Component {
  /**
   * @description Initilize the component.
   * @param {*} props for component initilization.
   */
  constructor (props) {
    super (props);
    this.state = {
      listOpen: false,
      headerTitle: this.props.title,
      filtered: false,
      mainList: this.props.list,
      list: this.props.list,
    };
  }

  /**
   * @description when component did update the add and remove event listener
   * based on dropdown if list is open or not
   */
  componentDidUpdate () {
    const {listOpen} = this.state;
    setTimeout (() => {
      if (listOpen) {
        window.addEventListener ('click', this.close);
      } else {
        window.removeEventListener ('click', this.close);
      }
    }, 0);
  }

  /**
     * @description When user will search the item in search box.
     * @param {Object} event The current taget object.
     */
  handleChange = event => {
    // Variable to hold the original version and filtered version of the list
    let currentList = [], newList = [];
    // If the search bar isn't empty
    if (event.target.value !== '') {
      this.setState ({
        filtered: true,
      });
      // Assign the original list to currentList
      currentList = this.state.list;
      // Use .filter() to determine which items should be displayed
      // based on the search terms
      newList = currentList.filter (item => {
        const lc = item.title.toLowerCase ();
        const filter = event.target.value.toLowerCase ();
        return lc.includes (filter);
      });
    } else {
      this.setState ({
        filtered: false,
      });
      // If the search bar is empty, set newList to original task list
      newList = this.state.mainList;
    }
    // Set the filtered state based on what our rules added to newList
    this.setState ({
      list: newList,
    });
  };

  /**
 * @description When user will click the check box based on user action toggle the item. *
 * @param {String} key key of list. if user is searching and item is slected the update both the list.
 * @param {Object} selectedList of selected items in dropdown.
 * @param {Object} mainList to set data in sync with filtered data.
 */
  toggleSelected = (key,selectedList,mainList) => {   
    this.setState (() => ({
        [key]: selectedList,
        mainList: mainList,
      }),() => {
        this.updateTitle ();
      }
    );
  };

  /**
   * @description Remove the click listener for dropdown.
   */
  componentWillUnmount () {
    window.removeEventListener ('click', this.close);
  }

  /**
   * @description close the open list when user will click for dropdown
   */
  close = () => {
    this.setState ({
      listOpen: false,
    });
  };

  /**
   * @description Apply condtions on title check whether the games are selected or not,
   *  and based on that change the title of dropdown
   */
  applyConditionsOnTitle () {
    const count = this.state.list.filter (function (a) {
      return a.selected;
    }).length;
    if (count === 0) {
      return {headerTitle: this.props.title};
    } else if (count === 1) {
      return {headerTitle: `${count} ${this.props.titleHelper}`};
    } else if (count > 1) {
      return {headerTitle: `${count} ${this.props.titleHelper}s`};
    }
  }

  /**
   * @description update Title of dropdown when user will select the ganmes.
   */
  updateTitle = () => {
    let obj = this.applyConditionsOnTitle ();
    this.setState ({
      headerTitle: obj.headerTitle,
    });
  };

  /**
   * @description Toggle the list when user will click on dropdown
   * based on user action
   */
  toggleList () {
    this.setState (prevState => ({
      listOpen: !prevState.listOpen,
    }));
  }

  /**
   * @description Render the dropdown component
   */
  render () {
    const {list} = this.state;
    const {mainList} = this.state;
    const {filtered} = this.state;
    const {options} = this.props;
    const {listOpen, headerTitle} = this.state;
    const containerStyle = {
      width: this.props.options[0].width,
      marginLeft: this.props.options[0].marginLeft,
    };
    return (
      <div style={containerStyle}>
        <div className="dd-wrapper">
          <div className="dd-header" onClick={() => this.toggleList()}>
            <div className="dd-header-title">{headerTitle}</div>
            {listOpen
              ? <FontAwesome name="angle-up" size="2x" />
              : <FontAwesome name="angle-down" size="2x" />}
          </div>
          {listOpen &&
            <ul className="dd-list" onClick={e => e.stopPropagation ()}>
              <Searchfilter
                handleChange={this.handleChange}
                options={options}
              />
              {list.map ((item, index) => (
                <div key={index} className="set-margin">
                  <CheckboxMultipleSelect
                    item={item}
                    toggleItem={this.toggleSelected}
                    updateTitles={this.updateTitles}
                    list={list}
                    mainList= {mainList}
                    filtered= {filtered}
                  />                 
                </div>
              ))}
            </ul>}
        </div>
      </div>
    );
  }
}

export default DropdownMultiple;
