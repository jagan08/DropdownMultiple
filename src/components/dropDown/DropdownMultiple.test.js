import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import DropdownMultiple from './DropdownMultiple';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

let options = [{
  width: '100%',
  placeHolderForSearchFilter: "foo..."
}];
let list= [{
    id: 0,
    title: 'Jacks or Better Double up',
    selected: false,
    key: 'list'
},
{
    id: 1,
    title: 'Creature from the black Lagoon',
    selected: false,
    key: 'list'
},
{
    id: 2,
    title: 'Go bananas',
    selected: false,
    key: 'list'
},
{
    id: 3,
    title: 'Ghost pirates',
    selected: false,
    key: 'list'
},
{
    id: 4,
    title: 'Berryburst',
    selected: false,
    key: 'list'
},
{
    id: 5,
    title: 'Butterfly stax',
    selected: false,
    key: 'list'
},
{
    id: 6,
    title: 'Arabian Nights',
    selected: false,
    key: 'list'
},
{
    id: 7,
    title: 'Excalibur',
    selected: false,
    key: 'list'
}
];

describe("DropdownMultiple Componenet", () => {
  test("it should render the app", () => {
    const wrapper = shallow(<DropdownMultiple options={options} />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  })

})

describe("applyConditionsOnTitle", () => {
  test("check the ttile based on selected items if items are not selected", () => {
    const wrapper = shallow(<DropdownMultiple options={options}  title="foo"/>);
    wrapper.setState({ list: list });
    const instance = wrapper.instance();
    const titleValue = instance.applyConditionsOnTitle();
    expect(titleValue.headerTitle).toBe('foo');
  });

  test("check the ttile based on selected items if 1 item is selected ", () => {
    let list= [{
      id: 0,
      title: 'Jacks or Better Double up',
      selected: true,
      key: 'list'
  }];
    const wrapper = shallow(<DropdownMultiple options={options}  title="foo" titleHelper="item"/>);
    wrapper.setState({ list: list });
    const instance = wrapper.instance();
    const titleValue = instance.applyConditionsOnTitle();
    expect(titleValue.headerTitle).toBe('1 item');
  });

  test("check the ttile based on selected items if more than 1 item is selected ", () => {
    let list= [{
      id: 0,
      title: 'Jacks',
      selected: true,
      key: 'list'
  },
  {
    id: 1,
    title: 'foo',
    selected: true,
    key: 'list'
}];
    const wrapper = shallow(<DropdownMultiple options={options}  title="foo" titleHelper="item"/>);
    wrapper.setState({ list: list });
    const instance = wrapper.instance();
    const titleValue = instance.applyConditionsOnTitle();
    expect(titleValue.headerTitle).toBe('2 items');
  });
})
