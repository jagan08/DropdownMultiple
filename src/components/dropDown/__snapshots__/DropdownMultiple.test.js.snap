// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMultiple Componenet it should render the app 1`] = `ShallowWrapper {}`;
